lister.cc

```diff
@@ -75,3 +77,44 @@ future<> lister::rmdir(fs::path dir) {
         return remove_file(dir.native());
     });
 }
+
+future<std::optional<directory_entry>> queue_lister::get() {
```

Comment:
I need to test that.
Doesn't sound too bad :)

Reply:
pipe didn't work well for me.
it isn't really abortable so it's a nightmare stopping in the middle of the directory listing.
you always need to consume all entries.

Reply:
What makes it not abortable? Could we fix pipe - or something like it - to be abortable?
Anyway, this was just an idea. It's not a must.

Reply:
> What makes it not abortable? Could we fix pipe - or something like it - to be abortable?
> Anyway, this was just an idea. It's not a must.

There's no function to close either the writer or the reader. Destroying the pipe while there's a blocking reader is fragile since it is synchronous and doesn't wait for the read to actually pop the "eof" marker.

I think this could be potentially fixed, but out of scope.
Just using the queue is simple enough and I don't see what value a pipe brings here.

Reply:
> There's no function to close either the writer or the reader. Destroying the pipe while there's a blocking reader is fragile since it is synchronous and doesn't wait for the read to actually pop the "eof" marker.

This was a design goal of pipes - that you have two **uncoupled** ends, the reader and writer, that can destroy their ends separately (there is no overall "pipe" object that you need to worry about). When the writer destroys his end, he can be assured that the reader will be woken up and read the eof - but can't wait for this to happen and doesn't want to wait for it to happen. It's usually completely pointless to wait for the reader to receive the eof - because, why wait to receive the eof, and not wait for - say - the reader to **handle** the eof?

So this decoupled ends is a feature of pipes, not something we'll want to fix. Maybe it's something we should even learn from: Yes, it's easy and sounds very responsible to have a `close()` method on everything and force the users to use it, but sometimes, we can invent simpler APIs that don't need an explicit `close()`.

Anyway, never mind, we don't have to use pipes :-)